src/agentscope/strategy/mixture_of_agent.py

```diff
@@ -0,0 +1,172 @@
+        messages = []
+        messages.append(
+            Msg(role="system", content=self.aggregator_prompt, name="system"),
+        )
+        for i, ref in enumerate(self.references, start=0):
+            messages.append(
+                Msg(
+                    role="user",
+                    content=ref,
+                    name=f"Model_{i}",
+                ),
+            )
+        aggre_format_msg = aggre_model.format(messages)
+        aggre_res = aggre_model(aggre_format_msg)
+        return aggre_res.text
+
+    def __call__(
+        self,
+        *args: Union[Msg, Sequence[Msg]],
+    ) -> str:
+        """
+        Get model response from messages.
+        Is equivalent to calling a model with:
+            ```
+            format_msg = model.format(messages)
+            return model(format_msg)
+            ```
+
+        Args:
+            *args (`Union[Msg, Sequence[Msg]]`):
+                The messages to be sent to the model.
+        """
+
+        def _process_reference(
+            i: int,
+            ref_model: ModelWrapperBase,
+            *args: Union[Msg, Sequence[Msg]],
+        ) -> Tuple[int, str]:
+            format_msg = ref_model.format(*args)
+            ref_model_res = ref_model(format_msg)
+            return i, ref_model_res.text
+
+        def _process_new_refs(
+            i: int,
+            ref_model: ModelWrapperBase,
+        ) -> Tuple[int, str]:
+            return i, self._get_res_with_aggregate_model(ref_model)
+
+        # get all the references
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            futures = [
+                executor.submit(_process_reference, i, ref_model, *args)
+                for i, ref_model in enumerate(self.reference_models, start=0)
+            ]
+            for future in concurrent.futures.as_completed(futures):
+                i, result = future.result()
+                self.references[i] = result
+                if self.show_internal:
+                    logger.info(f"Round {0}, Model_{i}: {result}")
+
+        for r in range(self.rounds):
+            if self.show_internal:
+                logger.info("=" * 20)
+            new_refs = ["" for _ in range(len(self.reference_models))]
+            with concurrent.futures.ThreadPoolExecutor() as executor:
+                futures = [
+                    executor.submit(_process_new_refs, i, ref_model)
+                    for i, ref_model in enumerate(
+                        self.reference_models,
+                        start=0,
+                    )
+                ]
+                for future in concurrent.futures.as_completed(futures):
+                    i, result = future.result()
+                    new_refs[i] = result
+                    if self.show_internal:
+                        print(f"Round {r+1}, Model_{i}: {result}")
+            self.references = new_refs
+
+        final_res = self._get_res_with_aggregate_model(self.main_model)
```

Comment:
Why dose `_get_res_with_aggregate_model` requires `self.main_model` as a input parameter (as `self.main_model` should be accessible in `_get_res_with_aggregate_model`)?

Reply:
Because in other rounds (from round 1 to round r-1), all the reference model is used as aggregate_model. So I pass the model as argument to reuse the `_get_res_with_aggregate_model` function.